aten/src/ATen/native/cuda/GridSampler.cu

```diff
@@ -57,8 +60,8 @@ namespace {
       scalar_t x = grid.data[grid_offset];
       scalar_t y = grid.data[grid_offset + grid_sCoor];
 
-      scalar_t ix = grid_sampler_compute_source_index(x, inp_W, padding_mode, align_corners);
-      scalar_t iy = grid_sampler_compute_source_index(y, inp_H, padding_mode, align_corners);
+      opmath_t ix = grid_sampler_compute_source_index(x, inp_W, padding_mode, align_corners);
```

Comment:
you have to cast `x` to `opmath_t` if you want higher precision result, otherwise `grid_sampler_compute_source_index` would run in `scalar_t`, and just upcast low-precision result.

Reply:
Sure. I have tried to promote `x`  to `opmath_t` on  `grid_sampler_compute_source_index`, but that would cause unit test failure on `test_grid_sample_large, issue_24823_1, torch.half`.

Specifically, in that test, `affine_grid` was done in half precision and created an output of (half) `x = 0.666503906250`. By calculating
```
((coord + 1.f) * size - 1) / 2
```
of `grid_sampler_unnormalize` in half precision we get exactly 2.0, but 1.999755859375 in float precision. With the latter value, it falls "within in bound" of [0, 2) and yields wrong results.

Since using non-promoted `scalar_t` dtype in `grid_sampler_compute_source_index` still yields the accurate result, as shown in the added unit test with half precision, I think it's ok to use the non-promoted type in this step. A more comprehensive solution would be to use promoted type in `affine_grid` as well, but that would take extra effort and beyond the scope of this PR's fix. What do you think?

Here is a short CUDA program that demonstrates this
```cuda
#include <cuda.h>
#include <cuda_runtime.h>
#include <cuda_fp16.h>

#include <iostream>

enum class GridSamplerPadding {Zeros, Border, Reflection};

template <typename scalar_t>
static __forceinline__ __device__
scalar_t grid_sampler_unnormalize(scalar_t coord, int size, bool align_corners) {
  // unnormalize coord from [-1, 1] to [-0.5, size - 0.5]
  return (((float)coord + 1.f) * size - 1) / 2;
}

// Computes the pixel source index value for a grid coordinate
template <typename scalar_t>
static __forceinline__ __device__
scalar_t grid_sampler_compute_source_index(
    scalar_t coord,
    int size,
    GridSamplerPadding padding_mode,
    bool align_corners) {
  printf("coord 1 = %.12f\n", (float)coord);
  coord = grid_sampler_unnormalize(coord, size, align_corners);
  printf("coord 2 = %.12f\n", (float)coord);
  return coord;
}

template <typename scalar_t>
__global__ void kernel(scalar_t* coord) {
  int size = 3;
  GridSamplerPadding padding_mode = GridSamplerPadding::Zeros;
  bool align_corners = false;

  scalar_t coord_out = grid_sampler_compute_source_index(*coord, size, padding_mode, align_corners);
  printf("coord_out = %.12f\n", (float)coord_out);
}

#ifndef TYPE_T
  #define TYPE_T float
#endif

int main() {
  // using type_t = __half;
  // using type_t = float;
  using type_t = TYPE_T;

  type_t h_coord = 0.666503906250;
  type_t *d_coord;
  cudaMalloc(&d_coord, sizeof(type_t) * 1);
  cudaMemcpy(d_coord, &h_coord, sizeof(type_t) * 1, cudaMemcpyHostToDevice);

  kernel<<<1, 1>>>(d_coord);
  cudaDeviceSynchronize();
}
```

Reply:
So in this particular case lower precision actually passes the test due to some compounding of errors? I'm pretty sure it will be different for other inputs, but ok we can leave it like this to unblock